quickwit-search/src/root.rs

```diff
@@ -0,0 +1,376 @@
+
+    // Find the sum of the number of hits and merge multiple partial hits into a single partial hits.
+    let mut leaf_search_results = Vec::new();
+    for leaf_search_response in leaf_search_responses {
+        match leaf_search_response {
+            Ok(leaf_search_result) => leaf_search_results.push(leaf_search_result),
+            Err(err) => error!(err=?err),
+        }
+    }
+    let leaf_search_result = spawn_blocking(move || collector.merge_fruits(leaf_search_results))
+        .await
+        .with_context(|| "Failed to merge leaf search results.")??;
+
+    // Create a hash map of PartialHit with split as a key.
+    let mut partial_hits_map: HashMap<String, Vec<PartialHit>> = HashMap::new();
+    for partial_hit in leaf_search_result.partial_hits.iter() {
+        partial_hits_map
+            .entry(partial_hit.split.clone())
+            .or_insert_with(Vec::new)
+            .push(partial_hit.clone());
+    }
+
+    // Perform the fetch docs phese.
+    let mut fetch_docs_handles: Vec<JoinHandle<anyhow::Result<FetchDocsResult>>> = Vec::new();
+    for (addr, jobs) in assigned_leaf_search_jobs.iter() {
+        for job in jobs {
+            let mut search_client = if let Some(client) = clients.get(&addr) {
+                client.clone()
+            } else {
+                return Err(anyhow::anyhow!(
+                    "There is no search client for {} in the search client pool.",
+                    addr
+                ));
+            };
+
+            // TODO: Change to a better way that does not use unwrap().
+            let partial_hits = partial_hits_map.get(&job.split).unwrap().clone();
+            let fetch_docs_request = FetchDocsRequest {
+                partial_hits,
+                index_id: search_request.index_id.clone(),
+            };
+            let handle = tokio::spawn(async move {
+                match search_client.fetch_docs(fetch_docs_request).await {
+                    Ok(resp) => Ok(resp.into_inner()),
+                    Err(err) => Err(anyhow::anyhow!("Failed to fetch docs due to {:?}", err)),
+                }
+            });
+            fetch_docs_handles.push(handle);
+        }
+    }
+    let fetch_docs_responses = futures::future::try_join_all(fetch_docs_handles).await?;
+
+    let mut hits: Vec<Hit> = Vec::new();
+    for response in fetch_docs_responses {
+        match response {
+            Ok(fetch_docs_result) => {
+                hits = merge_hits(&hits, &fetch_docs_result.hits)?;
+            }
+            Err(err) => error!(err=?err),
+        }
+    }
+
+    let elapsed = start_instant.elapsed();
+
+    Ok(SearchResult {
+        num_hits: leaf_search_result.num_hits,
+        hits,
+        elapsed_time_micros: elapsed.as_micros() as u64,
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use quickwit_proto::Hit;
+    use quickwit_proto::PartialHit;
+
+    use crate::root::merge_hits;
+
+    #[test]
+    fn test_root_merge_hits() {
```

Comment:
proptesting requires a bit of work and head scratching but is really brilliant for that kind of stuff.

We can try to do that this afternoon.